### PR TITLE
Avoid missing cube files on reading

### PR DESCRIPTION
**Pull request by osopardo1**
Due to the bug stated in issue #47 and according to reasoning in #48, we should work on avoiding losing cube records at reading time. 

The greedy estimation in the first step of the writing protocol causes having **missing parents** in the index structure. This means that all data is written down, but the read operation is not supposed to contemplate a behavior in which cube `"AAAA"` is present but parent `"AAA"` is not. If a father is not present in the index structure because of bad index quality, the query should be anyways retrieving all the necessary elements. 

For that, this Draft PR is for:

- [x] Use Patricia Trie structure instead of Map for resolving the CubeWeights. A Patricia Trie would find for any matches in a CubeId or its descendants. => No need, a SortedMap is enoguh
- [x] Possibly change the recursive implementation to an iterative one. This would help also with issue #56

### CODE REVIEW

**Comment by osopardo1** [comment]
Hey @Jiaweihu08 @cugni  you could check this one for the pair programming session today :smiling_imp:

**Review by cugni — CHANGES_REQUESTED** [review]

**Comment by cugni** [comment]
Why do we need the previouslyMappedFiles? What will delta filter that we cannot? Is there a specific case? if not, it would be better just to have the QbeastFiles directly to the indexStatus, instead of intersecting the two lists. Also, OTreeIndex, we are intersecting the file list twice, which is unnecessary.

File: src/main/scala/io/qbeast/spark/index/query/QueryIndexStatusExecutor.scala [hunk 1]
```diff
@@ -36,39 +39,88 @@ class QueryIndexStatusExecutor(querySpec: QuerySpec, indexStatus: IndexStatus) {
...
-            .filter(space.intersectsWith)
-            .flatMap(doFindSampleFiles)
-          if (!replicatedSet.contains(cube) && weightRange.from < maxWeight) {
-            val cubeFiles = files.flatMap(fileMap.get)
-            if (childFiles.nonEmpty) {
-              cubeFiles.filterNot(_.state == State.ANNOUNCED) ++ childFiles
-            } else {
-              cubeFiles
+
+    def doFindSampleFilesIterative(cube: CubeId): IISeq[QbeastFile] = {
```
[the remark this answers is not in the document]
Comment (eavilaes) [reply src/main/scala/io/qbeast/spark/index/query/QueryIndexStatusExecutor.scala]: Solved in 5a6623c

File: src/main/scala/io/qbeast/spark/index/query/QueryIndexStatusExecutor.scala [hunk 2]
```diff
@@ -5,6 +5,8 @@ package io.qbeast.spark.index.query
 
 import io.qbeast.IISeq
 import io.qbeast.core.model._
+import scala.collection.immutable.SortedMap
+import scala.collection.mutable.Queue
 import io.qbeast.spark.utils.State
 
 /**
```
[the remark this answers is not in the document]
Comment (eavilaes) [reply src/main/scala/io/qbeast/spark/index/query/QueryIndexStatusExecutor.scala]: Solved in 5a6623c

**Review by Jiaweihu08 — APPROVED** [review]

**Review by cugni — APPROVED** [review]
It seems good!

File: src/test/scala/io/qbeast/spark/index/FailingTests.scala [hunk 3]
```diff
@@ -0,0 +1,129 @@
+package io.qbeast.spark.index
+
+import io.qbeast.TestClasses.{Client3, T1, T2, TestStrings}
+import io.qbeast.core.model.{IndexStatus, QTableID}
+import io.qbeast.spark.QbeastIntegrationTestSpec
+import org.apache.spark.sql.{Dataset, SparkSession}
+
+class FailingTests extends QbeastIntegrationTestSpec with IndexTestChecks {
```
Comment (cugni) [inline src/test/scala/io/qbeast/spark/index/FailingTests.scala:8]: These tests need to be renamed.

**Review by cugni — CHANGES_REQUESTED** [review]
We still need to ensure that the QueryExecutor skips the files with maxWeight < weightRange.from

File: src/test/scala/io/qbeast/spark/index/FailingTests.scala (hunk 3 above)
Comment (eavilaes) [reply src/test/scala/io/qbeast/spark/index/FailingTests.scala]: The tests were already in `TransformerIndexingTest`, the only difference is that in `FailingTests` they had a small `cubeSize` set.
In 8cc6a4b I've added the. smallCubeSize test to `IndexTest`, but only for a base case; I don't think we need to test indexing different data types in this case.

**Review by eavilaes — COMMENTED** [review]
When creating a new test in `QueryExecutorTest.scala`, I found out that most of these tests are using a `cubeSize = 10`.
This leads to the `cubeSize` being ignored and the `config.DEFAULT_CUBE_SIZE` is used.
Is this the desired behavior for these tests?

**Review by cugni — APPROVED** [review]

**Pull request merged**